### PR TITLE
support Protobuf v2

### PR DESCRIPTION
**Pull request by Boyce-Lee**
Change-Id: Iac9cf713582236c303620d6f3c4c74cb0cba2303

### CODE REVIEW

**Comment by teverett** [comment]
@Boyce-Lee  could you add this into the root pom.xml please?

**Comment by Boyce-Lee** [comment]
>

**Comment by Boyce-Lee** [comment]
> @Boyce-Lee could you add this into the root pom.xml please?

done. 
thx for reviewing againt~

**Comment by teverett** [comment]
@Boyce-Lee  there is a problem with the pom.xml

```
The project  (/home/runner/work/grammars-v4/grammars-v4/protobuf2/pom.xml) has 1 error
Error:      Non-readable POM /home/runner/work/grammars-v4/grammars-v4/protobuf2/pom.xml: no more data available - expected end tag </project> to close start tag <project> from line 1, parser stopped on END_TAG seen ...</plugins>\n\t</build>\n... @56:1
```

**Comment by Boyce-Lee** [comment]
> @Boyce-Lee there is a problem with the pom.xml
> 
> ```
> The project  (/home/runner/work/grammars-v4/grammars-v4/protobuf2/pom.xml) has 1 error
> Error:      Non-readable POM /home/runner/work/grammars-v4/grammars-v4/protobuf2/pom.xml: no more data available - expected end tag </project> to close start tag <project> from line 1, parser stopped on END_TAG seen ...</plugins>\n\t</build>\n... @56:1
> ```

fixed

File: protobuf2/Protobuf2.g4 [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+
+// Option
+
+optionStatement
+  : OPTION optionName EQ constant SEMI
+  ;
+
+optionName
+  : fullIdent
+  | ( ident | ( LP fullIdent RP ) ) ( DOT fullIdent )?
```
Comment (KvanTTT) [inline protobuf2/Protobuf2.g4:50]: Redundant parentheses:

```suggestion
  | ( ident | LP fullIdent RP ) ( DOT fullIdent )?
```

File: protobuf2/Protobuf2.g4 [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+  ;
+
+enumElement
+  : optionStatement
+  | enumField
+  | emptyStatement_
+  ;
+
+enumField
+  : ident EQ ( MINUS )? intLit enumValueOptions?SEMI
```
Comment (KvanTTT) [inline protobuf2/Protobuf2.g4:184]:
```suggestion
  : ident EQ MINUS? intLit enumValueOptions?SEMI
```

File: protobuf2/Protobuf2.g4 [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+serviceElement
+  : optionStatement
+  | rpc
+  | stream
+  | emptyStatement_
+  ;
+
+rpc
+  : RPC rpcName LP ( STREAM )? messageType RP
+        RETURNS LP ( STREAM )? messageType RP
```
Comment (KvanTTT) [inline protobuf2/Protobuf2.g4:246]:
```suggestion
  : RPC rpcName LP STREAM? messageType RP
        RETURNS LP STREAM? messageType RP
```

File: protobuf2/Protobuf2.g4 [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+enumName: ident;
+fieldName: ident;
+groupName: ident;
+oneofName: ident;
+mapName: ident;
+serviceName: ident;
+rpcName: ident;
+streamName: ident;
+messageType: ( DOT )? ( ident DOT )* messageName;
+enumType: ( DOT )? ( ident DOT )* enumName;
```
Comment (KvanTTT) [inline protobuf2/Protobuf2.g4:287]:
```suggestion
messageType: DOT? ( ident DOT )* messageName;
enumType: DOT? ( ident DOT )* enumName;
```

File: protobuf2/Protobuf2.g4 [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+RC: '}';
+LT: '<';
+GT: '>';
+DOT: '.';
+COMMA: ',';
+COLON: ':';
+PLUS: '+';
+MINUS: '-';
+
+STR_LIT: ( '\'' ( CHAR_VALUE )*? '\'' ) |  ( '"' ( CHAR_VALUE )*? '"' );
```
Comment (KvanTTT) [inline protobuf2/Protobuf2.g4:355]:
```suggestion
STR_LIT: '\'' CHAR_VALUE*? '\'' |  '"' CHAR_VALUE*? '"';
```

File: protobuf2/Protobuf2.g4 [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+fragment CHAR_VALUE: HEX_ESCAPE | OCT_ESCAPE | CHAR_ESCAPE | ~[\u0000\n\\];
+fragment HEX_ESCAPE: '\\' ( 'x' | 'X' ) HEX_DIGIT HEX_DIGIT;
+fragment OCT_ESCAPE: '\\' OCTAL_DIGIT OCTAL_DIGIT OCTAL_DIGIT;
+fragment CHAR_ESCAPE: '\\' ( 'a' | 'b' | 'f' | 'n' | 'r' | 't' | 'v' | '\\' | '\'' | '"' );
+
+BOOL_LIT: 'true' | 'false';
+
+FLOAT_LIT : ( DECIMALS DOT DECIMALS? EXPONENT? | DECIMALS EXPONENT | DOT DECIMALS EXPONENT? ) | 'inf' | 'nan';
```
Comment (KvanTTT) [inline protobuf2/Protobuf2.g4:363]:
```suggestion
FLOAT_LIT : DECIMALS DOT DECIMALS? EXPONENT? | DECIMALS EXPONENT | DOT DECIMALS EXPONENT? | 'inf' | 'nan';
```

File: protobuf2/Protobuf2.g4 [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+fragment EXPONENT  : ( 'e' | 'E' ) (PLUS | MINUS)? DECIMALS;
+fragment DECIMALS  : DECIMAL_DIGIT+;
+
+INT_LIT     : DECIMAL_LIT | OCTAL_LIT | HEX_LIT;
+fragment DECIMAL_LIT : ( [1-9] ) DECIMAL_DIGIT*;
```
Comment (KvanTTT) [inline protobuf2/Protobuf2.g4:368]:
```suggestion
fragment DECIMAL_LIT : [1-9] DECIMAL_DIGIT*;
```

**Review by KvanTTT — COMMENTED** [review]
Remove some useless parentheses for clarity.

**Comment by KvanTTT** [comment]
Thank you for the contribution and welcome to ANTLR grammars!

**Comment by KvanTTT** [comment]
fixes #2583

**Pull request merged**